### PR TITLE
update units tutorial for thoroughness

### PR DESCRIPTION
**Pull request by Oblarg**
Previous units tutorial omitted a lot of important information; this improves the thoroughness.

### CODE REVIEW

File: source/docs/software/basic-programming/cpp-units.rst [hunk 1]
```diff
@@ -1,79 +1,116 @@
...
 
+Unit Types and Container Types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The C++ units library is based around two sorts of type definitions: unit types and container types.
+
+Unit Types
+~~~~~~~~~~
+
+Unit types correspond to the abstract concept of a unit, without any actual stored value.  Unit types are the fundamental "building block" of the units library - all units types are defined constructively (using the ``compound_unit`` template) from a small number of "basic" unit types (such as ``meters``, ``seconds``, etc).
```
Comment (prateekma) [inline source/docs/software/basic-programming/cpp-units.rst]: Should be "all unit types".

File: source/docs/software/basic-programming/cpp-units.rst [hunk 2]
```diff
@@ -1,79 +1,116 @@
...
+While unit types cannot contain numerical values, their use in building other unit types means that when a type or method uses a `template parameter <http://www.cplusplus.com/doc/oldtutorial/templates/>`__ to specify its dimensionality, that parameter will be a unit type.
+
+Container Types
+~~~~~~~~~~~~~~~
+
+Container types correspond to an actual quantity dimensioned according to some unit - that is, they are what actually hold the numerical value. Container types are constructed from unit types with the ``unit_t`` template.  Most unit types have a corresponding container type that has the same name suffixed by ``_t`` - for example, the unit type ``units::meter`` corresponds to the container type ``units::meter_t``.
+
+Whenever a specific quantity of a unit is used (as a variable or a method parameter), it will be an instance of the container type.  By default, container types will store the actual value as a ``double`` - advanced users may change this by calling the ``unit_t`` template manually.
+
+A full list of unit and container types can be found in the `documentation <http://nholthaus.github.io/units/namespaces.html>`__
```
Comment (prateekma) [inline source/docs/software/basic-programming/cpp-units.rst]: Needs period at end of sentence.

File: source/docs/software/basic-programming/cpp-units.rst [hunk 3]
```diff
@@ -1,79 +1,116 @@
...
+
+To create an instance of a specific unit, we create an instance of its container type:
 
 .. code-block:: c++
 
    // The variable speed has a value of 5 meters per second.
    units::meter_per_second_t speed{5.0};
 
-Alternatively, automatic type deduction using ``auto`` and literals can be used. This method is less verbose and recommended.
+Alternatively, the units library has `type literals <https://en.cppreference.com/w/cpp/language/user_literal>`__ defined for some of the more common container types.  These can be used in conjunction with type inference via ``auto`` to define a unit more-succinctly:
```
Comment (prateekma) [inline source/docs/software/basic-programming/cpp-units.rst]: Replace hyphen in "more-succinctly" with space.

File: source/docs/software/basic-programming/cpp-units.rst [hunk 4]
```diff
@@ -1,79 +1,116 @@
...
+   // Divide a meter_t value by a second_t, result is a meter_per_second_t
+   units::meter_per_second_t speed = 6_m / 0.5_s; // speed is 12_mps
+
+``<cmath>`` Functions
+^^^^^^^^^^^^^^^^^^^^^
+
+Some ``std`` functions (such as ``clamp``) are templated to accept any type on which the arithmetic operations can be performed.  Quantities stored as container types will work with these functions without issue.
 
-   units::meter_per_second_t speed = distance / time; // speed is 12_mps
+However, other ``std`` functions work only on ordinary numerical types (e.g. ``double``).  The units library's ``units::math`` namespace contains wrappers for several of these functions that accept units. Examples of such functions include ``sqrt``, ``pow``, etc.
```
Comment (prateekma) [inline source/docs/software/basic-programming/cpp-units.rst]: Do you think it's worth mentioning that `units::math::pow` is templated on the power, as opposed to the std implementation which is not?

Comment (Oblarg) [reply source/docs/software/basic-programming/cpp-units.rst]: Eh, not really.  If users are using that, they're probably deep enough to want to read the actual documentation.

Comment (prateekma) [reply source/docs/software/basic-programming/cpp-units.rst]: Fair enough.

File: source/docs/software/basic-programming/cpp-units.rst [hunk 5]
```diff
@@ -1,79 +1,116 @@
...
+While unit types cannot contain numerical values, their use in building other unit types means that when a type or method uses a `template parameter <http://www.cplusplus.com/doc/oldtutorial/templates/>`__ to specify its dimensionality, that parameter will be a unit type.
+
+Container Types
+~~~~~~~~~~~~~~~
+
+Container types correspond to an actual quantity dimensioned according to some unit - that is, they are what actually hold the numerical value. Container types are constructed from unit types with the ``unit_t`` template.  Most unit types have a corresponding container type that has the same name suffixed by ``_t`` - for example, the unit type ``units::meter`` corresponds to the container type ``units::meter_t``.
+
+Whenever a specific quantity of a unit is used (as a variable or a method parameter), it will be an instance of the container type.  By default, container types will store the actual value as a ``double`` - advanced users may change this by calling the ``unit_t`` template manually.
+
+A full list of unit and container types can be found in the `documentation <http://nholthaus.github.io/units/namespaces.html>`__.
```
Comment (AustinShalit) [inline source/docs/software/basic-programming/cpp-units.rst]:
```suggestion
A full list of unit and container types can be found in the `documentation <https://nholthaus.github.io/units/namespaces.html>`__.
```

**Pull request merged**